File: src/ios/ChromaKeyCamera.swift
@objc(ChromaKeyCamera) class ChromaKeyCamera : CDVPlugin, ChromaKeyCameraDelegate {
    
  var latestCall : CDVInvokedUrlCommand?
  var ckcviewController : ChromaKeyCameraViewController?

  @objc(start:) func start(_ command: CDVInvokedUrlCommand) {
    latestCall = command

    // argument order
    //[mode, backgroundMode, backgroundPhoto, backgroundVideo, color, threshold, smoothing]
    let argMode = command.arguments[0] as? String ?? ""
    let argBackgroundMode = command.arguments[1] as? String ?? ""
    let argBackgroundPhoto = command.arguments[2] as? String ?? ""
    let argBackgroundVideo = command.arguments[3] as? String ?? ""
    let argColor = command.arguments[4] as? String ?? ""
    let argThreshold = command.arguments[5] as? NSNumber ?? 0.0
    let argSmoothing = command.arguments[6] as? NSNumber ?? 0.0
    
    
    UIDevice.current.setValue(UIInterfaceOrientation.landscapeRight.rawValue, forKey: "orientation")
    
    DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(100)) {
    
        self.ckcviewController = ChromaKeyCameraViewController()
        
        self.ckcviewController!.modalPresentationStyle = .fullScreen
    
        switch argMode {
        case "video":
            self.ckcviewController!.mode = .video
        case "photo":
            self.ckcviewController!.mode = .photo
        default:
            self.returnError(message: "invalid mode")
            return
        }
    
        switch argBackgroundMode {
        case "video":
            self.ckcviewController!.backgroundMode = .video
            self.ckcviewController!.backgroundVideoURL = argBackgroundVideo
        case "photo":
            self.ckcviewController!.backgroundMode = .photo
            self.ckcviewController!.backgroundPhotoURL = argBackgroundPhoto
        default:
            self.returnError(message: "invalid background mode")
            return
        }
    
        switch argColor {
        case "red":
            self.ckcviewController!.color = .red
        case "blue":
            self.ckcviewController!.color = .blue
        case "green":
            self.ckcviewController!.color = .green
        default:
            self.returnError(message: "invalid color")
            return
        }
    
        self.ckcviewController!.threshold = argThreshold.floatValue
        self.ckcviewController!.smoothing = argSmoothing.floatValue
    
        self.ckcviewController!.delegate = self

        self.viewController.present(self.ckcviewController!, animated: true, completion: nil)
    }

  }
    
  func returnError(message: String) {
    let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: message)
    self.commandDelegate!.send(pluginResult, callbackId: latestCall!.callbackId)
  }
    
  func error(message: String) {
    self.ckcviewController?.dismiss(animated: true, completion: nil)
    self.ckcviewController = nil
    returnError(message: message)
  }
  
  func success(path: String) {
    self.ckcviewController?.dismiss(animated: true, completion: nil)
    self.ckcviewController = nil
    let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: path)
    self.commandDelegate!.send(pluginResult, callbackId: latestCall!.callbackId)
  }
  
  func calibrationSuccess(threshold: Float, smoothing: Float) {
    self.ckcviewController?.dismiss(animated: true, completion: nil)
    self.ckcviewController = nil
  }
  
}

File: www/ChromaKeyCamera.js
var exec = require('cordova/exec');

function checkArg(arg, def) {
  if(arg) {
    return arg
  }
  if(def) {
    return def
  }
  throw(new Error('missing mandatory argument'))
}

// var mode : OperationMode = .video
// var backgroundMode : BackgroundMode = .video
// var backgroundPhotoURL : String = ""
// var backgroundVideoURL : String = ""
// var color : Color = .blue
// var threshold : Float = 0.4
// var smoothing : Float = 0.1

exports.start = function(arg, success, error) {
  try {
    var mode = checkArg(arg.mode, 'photo')
    var backgroundMode = checkArg(arg.backgroundMode, 'photo')
    var backgroundPhoto = checkArg(arg.backgroundPhoto, backgroundMode == 'video' ? 'not required' : undefined)
    var backgroundVideo = checkArg(arg.backgroundVideo, backgroundMode == 'photo' ? 'not required' : undefined)
    var color = checkArg(arg.color, 'blue')
    var threshold = checkArg(arg.threshold, 0.4)
    var smoothing = checkArg(arg.smoothing, 0.1)
  }
  catch (err) {
    error(err)
    return
  }
  exec(success, error, 'ChromaKeyCamera', 'start', [mode, backgroundMode, 
    backgroundPhoto, backgroundVideo, color, threshold, smoothing])
}
